### ISSUE TITLE
Compressed storage migration path

### ISSUE THREAD

**Issue by wchargin** [sourcecred/sourcecred#2768]
(Paired with: @topocount, @hammadj, @wchargin.)

## Motivation
Users are hitting GitHub file size limits when their `credResult.json` file is larger than 100 MB, the hard cap for GitHub's API.

Goal: unblock users from deploying to GitHub Pages even if files like `credResult.json` exceed 100 MB in size.

## Plan
We'll compress data within physical files to honor GitHub's per-file size limit. This should be a fairly surgical change, without requiring any new technologies or patterns.

Define two interfaces:

```javascript
/*
 * A byte string represented as a JavaScript string with
 * code points between 0 and 255, inclusive.
 */
type ByteString = string;

interface DataStorage {
  get(key: string): Promise<ByteString>;
  // may also want "list"
}

interface WritableDataStorage extends DataStorage {
  set(key: string): Promise<ByteString>;
  // may also want "delete", "delete by prefix", ...
}
```

* Implement DiskStorage (node only) that implements the `WriteableDataStorage` interface
* Update CLI to use `DiskStorage` instead of using `fs` directly
* Rewrite `ledgerDiskStorage` to `NetworkStorage` (implementing `WriteableDataStorage`, for arbitrary files rather than just the ledger)
* Change `LedgerManager` usage on frontend to use `NetworkStorage`
* Implement `ZipStorage`, which can wrap any `DataStorage` implementation to shard logical files across physical files.
* Replace usages of `DiskStorage` and `NetworkStorage` in the CLI and frontend with sharded equivalents.

## Non-goals, future work
We considered using Git LFS. It may well be a better long-term solution. For now, we don't want to incur the additional complexity for either us or our users.

As we shard files, we still write them to disk, and require manual pushes to GitHub by instance maintainers. Direct-to-remote writes are a whole separate can of worms.

The fact that we have to resort to sharding is a clear signal that we should actively investigate more sustainable storage solutions for the long term. Happily, we're already doing that.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200032326400385/1200038250955742) by [Unito](https://www.unito.io/learn-more)

**Comment by wchargin** [comment]
cc @decentralion; please review and LGTM.

cc @blueridger: advisory re: your API/IO changes; please feel free to
offer any feedback.

**Comment by topocount** [comment]
@wchargin the more reading through docs and existing code I've done, the more confused I've gotten about the need for a ByteString type. My intuition was that it was needed to make tallying the size of a string more straightforward, but there are other tools we can use to calculate the size of some segment of data using the blob and buffer/arraybuffer libraries if we want.

Is ByteString needed because of some transport-layer consideration then?

**Comment by wchargin** [comment]
Are you asking about why a `ByteString` type is needed, or what that
type should be?

A storage layer stores blobs, which are sequences of bytes. We have to
represent these blobs in JavaScript memory somehow, and that is the need
for the `ByteString` type. A byte string is whatever is passed to `set`
and returned from `get`, and it has to have _some_ type.

Which, then? In Node, we have the `Buffer` type, which would probably be
the most appropriate choice, but there is no `Buffer` in the browser.
And the browser has `Blob`, which, despite the name, isn’t really the
right abstraction (it’s designed for HTTP interactions, and so has
baggage like content types and encodings that we do not want); plus,
it’s not in Node, anyway.

So three reasonable options come to my mind:

 1. `string`s whose code points are all between 0 and 255, inclusive; or
 2. `number[]`s with no empty slots whose values are all between 0 and
    255, inclusive; or
 3. `Uint8Array`s.

These three types are isomorphic, so the choice among them boils down to
which is most useful in a JS context. While we were sketching the draft,
I picked the first option (restricted `string`s) for simplicity and to
keep us focused on the broader shape of the API. Really, I think that
`Uint8Array`s may well be the best choice, because they interoperate
with `TextEncoder` and can’t be misinterpreted as text strings.

**Comment by decentralion** [comment]
This approach will allow us to store files >100MB in size, at the expense of making the files no longer "just JSON" that's easily readable or inspectable without special interfaces. As such, analysis workflows like "just grab credResult.json and load it into python and then..." will break.

Another approach which would allow us to store larger files at the expense of not having raw JSON would be to gzip the files at rest. Empirically, gzip tends to be very effective on our json blobs, so this would get us a lot of room before we hit the 100MB file size limit. This would also make it a lot easier to do other analysis flows, e.g. in python. We have `pako` so we can decompress the gzip'd contents manually in the browser. Was this alternative considered?

**Comment by topocount** [comment]
> cc @blueridger: advisory re: your API/IO changes; please feel free to
> offer any feedback.
@wchargin I;ve rebased onto these changes in my DataStorage implementation

**Comment by topocount** [comment]
The merge order for these PRs is as follows:

1. #2776 : Implement DiskStorage and the DataStorage interfaces
2. #2805 : implement isomorphic Text Encoder / Decoder
2. #2777 : Utilize DiskStorage
3. #2840 : Generalize disk utils as storage utils for use in browser
4. #2844 : Implement and utilize Network Storage
5. #2816 : Implement ZipStorage
6. #XXXX : Utilize ZipStorage
